### PR TITLE
Fix old GCC ABI check in CMake package config

### PR DESCRIPTION
**Pull request by cbalioglu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72081

This PR fixes the libstdc++ ABI check in CMake package configuration file (i.e. `TorchConfig.cmake`) The `_GLIBCXX_USE_CXX11_ABI` flag is a property of `libstdc++`, not GNU compiler collection. In its current form C++ libraries built with Clang on Linux fail since the `torch` CMake target propagates `_GLIBCXX_USE_CXX11_ABI` only when used with gcc.

Differential Revision: [D33899849](https://our.internmc.facebook.com/intern/diff/D33899849/)

### CODE REVIEW

**Review by malfet — APPROVED** [review]

File: cmake/TorchConfig.cmake.in [hunk 1]
```diff
@@ -166,9 +166,11 @@ if(@USE_CUDA@)
   list(APPEND TORCH_LIBRARIES ${TORCH_CUDA_LIBRARIES})
 endif()
 
-# When we build libtorch with the old GCC ABI, dependent libraries must too.
-if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
+# When we build libtorch with the old libstdc++ ABI, dependent libraries must too.
+if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
   set(TORCH_CXX_FLAGS "-D_GLIBCXX_USE_CXX11_ABI=@GLIBCXX_USE_CXX11_ABI@")
+else()
+  set(TORCH_CXX_FLAGS "-D_GLIBCXX_USE_CXX11_ABI=0")
```
Comment (malfet) [inline cmake/TorchConfig.cmake.in:173]: Can you please elaborate why this is needed? This define should not do anything on non-Linux platforms, should it?

Comment (cbalioglu) [reply cmake/TorchConfig.cmake.in:173]: It shouldn't be needed. Just trying to figure out why most of the builds have started to fail after this change with a `error: invalid token at start of a preprocessor expression`.

Comment (malfet) [reply cmake/TorchConfig.cmake.in:173]: You probably just need to rebase your PR against commit from `viable/strict` (there were several regressions this morning)

Comment (cbalioglu) [reply cmake/TorchConfig.cmake.in]: Ok, resolved the issue. Apparently caffe2's CMakeLists.txt was the culprit and had a similar check that caused GLIBCXX_USE_CXX11_ABI  variable to become empty. Everything is green now.

**Pull request closed without merging**